### PR TITLE
improve file list comparison

### PR DESCRIPTION
**Pull request by timosachsenberg**
## Description

as requested by @jpfeuffer 

## Checklist
- [ ] Make sure that you are listed in the AUTHORS file
- [ ] Add relevant changes and new features to the CHANGELOG file
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] New and existing unit tests pass locally with my changes
- [ ] Updated or added python bindings for changed or new classes (Tick if no updates were necessary.)

### How can I get additional information on failed tests during CI
<details>
  <summary>Click to expand</summary>
If your PR is failing you can check out

- The details of the action statuses at the end of the PR or the "Checks" tab.
- http://cdash.openms.de/index.php?project=OpenMS and look for your PR. Use the "Show filters" capability on the top right to search for your PR number.
  If you click in the column that lists the failed tests you will get detailed error messages.

</details>

### Advanced commands (admins / reviewer only)
<details>
  <summary>Click to expand</summary>
  
- `/reformat` (experimental) applies the clang-format style changes as additional commit. Note: your branch must have a different name (e.g., yourrepo:feature/XYZ) than the receiving branch (e.g., OpenMS:develop). Otherwise, reformat fails to push.
- setting the label "NoJenkins" will skip tests for this PR on jenkins (saves resources e.g., on edits that do not affect tests)
- commenting with `rebuild jenkins` will retrigger Jenkins-based CI builds
  
</details>

---
:warning: Note: Once you opened a PR try to minimize the number of *pushes* to it as every push will trigger CI (automated builds and test) and is rather heavy on our infrastructure (e.g., if several pushes per day are performed).

___

### **PR Type**
Enhancement, Tests

___

### **Description**
- Refactored the `validateMatchingFileNames` function to return a `MatchingFileListsStatus` enum instead of a boolean, improving clarity and error handling.
- Introduced new status codes: `MATCH`, `ORDER_MISMATCH`, and `SET_MISMATCH` to represent different validation outcomes.
- Added comprehensive tests for the `validateMatchingFileNames` function to cover various scenarios, ensuring robustness.
- Updated the `ProteomicsLFQ` class to utilize the new validation logic, enhancing error reporting and handling.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>File.cpp</strong><dd><code>Refactor file list validation to use status codes</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/openms/source/SYSTEM/File.cpp

<li>Changed return type of <code>validateMatchingFileNames</code> to <br><code>MatchingFileListsStatus</code>.<br> <li> Removed <code>strict</code> parameter and adjusted logic for file list validation.<br> <li> Introduced new status codes for file list validation: <code>MATCH</code>, <br><code>ORDER_MISMATCH</code>, and <code>SET_MISMATCH</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7645/files#diff-e7d2d98e6bc31e659d7a9ed853b63d69a365b252c78fa83f0f51c8be4732cbec">+42/-39</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>ProteomicsLFQ.cpp</strong><dd><code>Update file validation logic in ProteomicsLFQ</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/topp/ProteomicsLFQ.cpp

<li>Updated calls to <code>validateMatchingFileNames</code> to handle new status codes.<br> <li> Improved error handling based on validation results.<br>

</details>

  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7645/files#diff-1a8349c94e787728b67eeb8af8476400a6fcfb06d8f845940ccf32b81c73f1f9">+16/-6</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>File.h</strong><dd><code>Define status codes for file list validation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/openms/include/OpenMS/SYSTEM/File.h

<li>Defined <code>MatchingFileListsStatus</code> enum for file list validation results.<br> <li> Updated <code>validateMatchingFileNames</code> function signature and <br>documentation.<br>

</details>

  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7645/files#diff-db9442289ec1f9a58a5dad7411cc13c2d928509b63c5ad43c8b30eed8357f3d1">+19/-10</a>&nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Tests</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>File_test.cpp</strong><dd><code>Add comprehensive tests for file list validation</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

src/tests/class_tests/openms/source/File_test.cpp

<li>Added new tests for <code>validateMatchingFileNames</code> covering different <br>scenarios.<br> <li> Tested scenarios include exact match, order mismatch, set mismatch, <br>and basename comparison.<br>

</details>

  </td>
  <td><a href="https://github.com/OpenMS/OpenMS/pull/7645/files#diff-a747f5274b83bc8b6410d27f1121915e6553d8835a41997f8bbaebc142c08d54">+73/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**: Comment `/help "your question"` on any pull request to receive relevant information

### CODE REVIEW

File: src/openms/include/OpenMS/SYSTEM/File.h [hunk 1]
```diff
@@ -287,27 +287,36 @@ namespace OpenMS
...
+      This helper function performs validation and returns one of three states:
+      - MATCH (0): Files match perfectly (considering basename/extension settings)
+      - ORDER_MISMATCH (1): Same set of files but in different order
+      - SET_MISMATCH (2): Different sets of files (including different counts)
 
-      Note: Because workflow systems may assign file names randomly a non-strict comparison mode is enabled by default.      
-      Instead of the strict comparison (which returns false if there is a single mismatch), the non-strict comparison mode 
-      only returns false if the unique set of filenames match but some positions differ, i.e., only the order has been mixed up.
+      Note: Because workflow systems may assign file names randomly, this function 
+      focuses on detecting order mismatches rather than exact content differences.
```
Comment (timosachsenberg) [inline src/openms/include/OpenMS/SYSTEM/File.h]:
```suggestion
```

File: src/tests/class_tests/openms/source/File_test.cpp [hunk 2]
```diff
@@ -350,6 +350,79 @@ START_SECTION(File::download(std::string url, std::string filename))
...
+    StringList list1 = {"/path/to/file1.txt", "/different/path/file2.txt"};
+    StringList list2 = {"/somewhere/file2.txt", "/other/path/file1.txt"};
+    TEST_TRUE(File::validateMatchingFileNames(list1, list2, true, false) ==  File::MatchingFileListsStatus::ORDER_MISMATCH)
+  }
+
+  // Test ignore extension
+  {
+    StringList list1 = {"file1.txt", "file2.mzML"};
+    StringList list2 = {"file1.mzML", "file2.txt"};
+    TEST_TRUE(File::validateMatchingFileNames(list1, list2, false, true) == File::MatchingFileListsStatus::ORDER_MISMATCH)
```
Comment (timosachsenberg) [inline src/tests/class_tests/openms/source/File_test.cpp:401]:
```suggestion
    TEST_TRUE(File::validateMatchingFileNames(list1, list2, false, true) == File::MatchingFileListsStatus::SET_MISMATCH)
```

File: src/topp/ProteomicsLFQ.cpp [hunk 3]
```diff
@@ -1275,13 +1320,22 @@ class ProteomicsLFQ :
       ++fraction_group;
     }
 
-    // Check for common mistake that order of input files have been switched.
-    // This is the case if basenames are identical but the order does not match.
-    if (!File::validateMatchingFileNames(in_MS_run, id_MS_run_ref, true, true, false)) // only basenames, without extension, only order
+    // validate file lists (use only basename and ignore extension)
+    auto validation_result = File::validateMatchingFileNames(in_MS_run, id_MS_run_ref, true, true);
```
Comment (timosachsenberg) [inline src/topp/ProteomicsLFQ.cpp:1324]: I think this is now stricter than before. Previously, if at least one file had a different name (but it was not just a simple reordering) we did not fail.
Hope this is fine.

Comment (jpfeuffer) [reply src/topp/ProteomicsLFQ.cpp:1324]: What do you mean by this? I think this is the reason I did not understand the previous function. When can this happen?

Comment (jpfeuffer) [reply src/topp/ProteomicsLFQ.cpp:1324]: If it has a different name and is not a reordering then it was a different set and we should fail, right?

Comment (timosachsenberg) [reply src/topp/ProteomicsLFQ.cpp:1324]: Depends on whether you are checking file names passed to plfq or if filenames are annotated in the idXML.
We check both.
For file names passed on the CLI we are less strict and allow, e.g., random names for id and mzMLs as long as we can later confirm that the reference in the idXML references existing mzML.
For references in the idXML, we were also less strict (probably to support idXMLs or mzIdentMLs from arbitrary sources).
Now we are also stricter here.

Comment (jpfeuffer) [reply src/topp/ProteomicsLFQ.cpp:1324]: I think these days we should definitely be strict about a matching idXML origin and mzML origin.

For the input file names, we should quickly check if this basename check is working with default naming in KNIME and potentially TOPPAS (for standard workflows). If it is compatible, we should only allow mismatching filenames if the advanced user specifically opts in.

Comment (timosachsenberg) [reply src/topp/ProteomicsLFQ.cpp:1324]: k then let's merge. I think we are planning to test the other KNIME fixes as well.
Note that I even added a test in the past that allows mismatching filenames on the CLI so I am pretty sure there are some use cases.

**Pull request merged**